ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java

```diff
@@ -1007,17 +1001,12 @@ protected T getXpathOrFuncExprNodeDesc(ASTNode node,
             T columnDesc = children.get(0);
             T valueDesc = interpretNode(columnDesc, children.get(i));
             if (valueDesc == null) {
-              if (hasNullValue) {
-                // Skip if null value has already been added
-                continue;
-              }
-              TypeInfo targetType = exprFactory.getTypeInfo(columnDesc);
+              // Keep original
+              TypeInfo targetType = exprFactory.getTypeInfo(children.get(i));
               if (!expressions.containsKey(targetType)) {
                 expressions.put(targetType, columnDesc);
               }
-              T nullConst = exprFactory.createConstantExpr(targetType, null);
-              expressions.put(targetType, nullConst);
-              hasNullValue = true;
+              expressions.put(targetType, children.get(i));
             } else {
```

Comment:
I was going thru here and there and I think there might be another way around this problem which could retain this optimization as well:
* introduce a new `NOT` operator: which can be controlled to return true/false in case of null values
* in case of filter expressions start using the new not operator; and switch mode below every `NOT` operator

but this feels like a more complicated change - we should only do it if we loose important optimizations

Reply:
other idea could be to retain and do these optimizations but only if we are not below a NOT operator

Reply:
So I'm not sure how to address your comments, I"m gonna address all of them here.  Lemme take a step back and tell you what I've already discussed with Jesus and the direction we were going.

So we know we're losing some optimizations as you've noted.  Jesus felt that they weren't that big of a deal.  For instance, we'd lose the optimization of "tinyint_col in (2500000)".  Previously, we saw this changing to false since a tinyint col can never be that value, but that check won't be optimized out now.  I think that's the main one I saw with the tests.

So for now, we'd like to avoid a bigger rewrite.  He also noted that this optimization should perhaps be more in the Calcite framework, which makes sense to me.

You did have one other comment about an "if" statement not being hit.  I'm not sure I understand the bug that you're referring to.  It's a bit complicated to understand, but it seems ok to me?  Can you explain this further?
Thanks again!

Reply:
yes they could be in Calcite some day...we can live without them

that if statement is easier to be viewed without all these comments:

https://github.com/apache/hive/blob/20137b5894fc7ec6f4acaef4304dfdc2ce93275a/ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java#L1012-L1015

Reply:
Oh, I see.  It is a multimap though, so I think it retains all values?

Reply:
Cloud we repace null with Boolean.False here in this case?  to expression like  false or col in (int1, int2,..)  or col in (...)

Reply:
I'm sorry...which statement are you referring to where you want to replace null with "false"?

Reply:
I mean the way we deal with the constant that cannot be interpreted, treat it as a null or false expression.

 For example:
```col not in (355.8)```  translates to ```not(null)```;
```col not in (355.8, 123)```  translates to ```not(null or col in (123))```;

Reply:
sorry,  not(null) seems do not work as expected....

Reply:
Yeah, not in (null) is a very weird construct.  Equivalent to why you need to say "x IS NULL" and you can't say "x = NULL"